### ISSUE TITLE
Add OCPL html filtering to logs/submit

### ISSUE THREAD

**Issue by following5** [opencaching/okapi#412]
See [comments](https://github.com/opencaching/okapi/blob/7e2e3796c7e2902be4f4ea268c36e960de3aeffb/okapi/services/logs/submit.php#L324) in submit code.

**Comment by wrygiel** [comment]
Just making sure: Have you checked your latest commit against XSS attacks?

**Comment by wrygiel** [comment]
Here's a comment on `cache_desc.desc_html` column in OCPL:

> Format in which `desc` column is encoded: 0 - DO NOT USE (unknown format based on HTML); 1 - unsafe HTML (needs to be purified before it is included on a HTML page); 2 - safe HTML (may be included "as is" on HTML pages, without any further processing).

This means, that if you save something with `desc_html=2` then you MUST clean it beforehand.

**Comment by following5** [comment]
No. But the only effective change is that I enabled this code ...

```
    $formatted_comment = str_replace("&amp;", "&amp;#38;", $formatted_comment);
    $formatted_comment = str_replace("&lt;", "&amp;#60;", $formatted_comment);
    $formatted_comment = str_replace("&gt;", "&amp;#62;", $formatted_comment);
```

for OCPL html comments; the rest was only restructuring.

**Comment by wrygiel** [comment]
> No. But the only effective change is that I enabled this code

Ok. Then it couldn't have broken anything.

**Comment by following5** [comment]
> Here's a comment on cache_desc.desc_html column in OCPL:

This issue is only about the `cache_logs.text` column.

**Comment by following5** [comment]
This is a serious problem. OKAPI is the only way to insert bad HTML code into OCPL databases, after HTML purifying has been added for all HTML input on the OCPL websites.

There is a class `userInputFilter` in OCPL code which should be included in OKAPI. But it includes loads of other OCPL code, just for one reason: To cache the HTML purifier settings. I am not sure how to proceed here.

**Comment by wrygiel** [comment]
Perhaps @f-ranek might be able to suggest something?

If I remember correctly, there were other places inserting bad HTML into OCPL databases. And the purification was done dynamically before the content from the database was displayed.

**Comment by following5** [comment]
>  the purification was done dynamically before the content from the database was displayed.

It should be done. But `services/logs/submit?comment_format=html&comment=<em>` will make _all_ logs of the geocache appear in italics. Or is there some cronjob which will clean it up?

**Comment by wrygiel** [comment]
> Or is there some cronjob which will clean it up?

I don't think so ;)

**Comment by following5** [comment]
Just noticed that comments in 'auto' format are not filtered even for OCDE. Must be filtered because it is interpreted as HTML.

**Comment by following5** [comment]
Note that the OCPL HTML filtering, which is now included in logs/submit, is not that sophisticated. Other than the the bullet-proof "Purifier" used by OCDE it will detect only some forms for bad HTML, not all.

**Comment by wrygiel** [comment]
For years now, it has been very easy to inject XSS in OCPL. For example, see [here](https://opencaching.pl/viewcache.php?cacheid=10492) (scroll to the bottom of the page, and my injected JavaScript fires up).

**Comment by following5** [comment]
> For years now, it has been very easy to inject XSS in OCPL. For example, see here (scroll to the bottom of the page, and my injected JavaScript fires up).

[fixed](https://github.com/opencaching/opencaching-pl/commit/59316bf7df808bbabd2542e96f47637207d21891) :-)

I will wait a few days, and then Okapi log submitting can be adjusted: set `text_html`=2 for new OCPL logs and discard the double-encoding of HTML entities.

(`text_html`=1 for OCPL now means "broken old logs", =2 means "clean new logs")

**Comment by following5** [comment]
> Note that the OCPL HTML filtering, which is now included in logs/submit, is not that sophisticated. Other than the the bullet-proof "Purifier" used by OCDE it will detect only some forms for bad HTML, not all.

Just noticed that there is also an HTMLPurifier setup in OCPL code: https://github.com/opencaching/opencaching-pl/blob/master/lib/userInputFilters/userInputFilter.php

This should be used instead of the old filter function.

**Comment by following5** [comment]
fixed by ec7515bf (thanks!)
